core/src/main/scala/com/metamx/tranquility/beam/Beam.scala

```diff
@@ -19,22 +19,37 @@
 package com.metamx.tranquility.beam
 
 import com.twitter.util.Future
+import scala.collection.immutable.BitSet
+import scala.collection.mutable
 
 /**
- * Beams can accept events and forward them along. The propagate method may throw a DefunctBeamException, which means
- * the beam should be discarded (after calling close()).
- */
-trait Beam[A]
+  * Beams can accept events and forward them along. The propagate method may throw a DefunctBeamException, which means
+  * the beam should be discarded (after calling close()).
+  */
+trait Beam[MessageType]
 {
   /**
-   * Request propagation of events. The operation may fail in various ways, which tend to be specific to
-   * the implementing class. Returns the number of events propagated.
-   */
-  def propagate(events: Seq[A]): Future[Int]
+    * Request propagation of events. The operation may fail in various ways, which tend to be specific to
+    * the implementing class.
+    * @param messages a batch of messages
+    * @return the number of events propagated
+    */
+  @deprecated("use sendBatch", "0.7.0")
+  final def propagate(messages: Seq[MessageType]): Future[Int] = {
+    sendBatch(messages).map(_.size)
+  }
+
+  /**
+    * Request propagation of events. The operation may fail in various ways, which tend to be specific to
+    * the implementing class.
+    * @param messages a batch of messages
+    * @return a bitset containing indexes of messages that sent successfully; other messages were dropped.
+    */
+  def sendBatch(messages: Seq[MessageType]): Future[BitSet]
```

Comment:
if indicies are returned, parameter should probably be `scala.collection.immutable.Seq`